bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java

```diff
@@ -0,0 +1,213 @@
+        }
+    }
+
+    private void disconnect() {
+        logger.debug("Disconnect from device {}...", address);
+        if (!device.disconnect()) {
+            logger.debug("Disconnect from device {} failed", address);
+        }
+    }
+
+    private void discoverServices() {
+        logger.debug("Discover services for device {}", address);
+        if (!device.discoverServices()) {
+            logger.debug("Discovering services failed");
+            disconnect();
+        }
+    }
+
+    private void read() {
+        logger.debug("Read data from device {}...", address);
+        BluetoothCharacteristic characteristic = device.getCharacteristic(uuid);
+        if (!device.readCharacteristic(characteristic)) {
+            logger.debug("Read data from device {} failed", address);
+            disconnect();
+        }
+    }
+
+    private void printServices() {
+        device.getServices().forEach(service -> logger.debug("Device {} Service '{}'", address, service));
+    }
+
+    @Override
+    public void onServicesDiscovered() {
+        synchronized (servicesResolved) {
+            servicesResolved = true;
+        }
+        logger.debug("Service discovery completed for device {}", address);
+        printServices();
+        execute();
+    }
+
+    @Override
+    public void onConnectionStateChange(BluetoothConnectionStatusNotification connectionNotification) {
+        switch (connectionNotification.getConnectionState()) {
+            case DISCOVERED:
+                logger.debug("Device {} DISCOVERED", address);
+                startScheduledTask();
+                break;
+            case CONNECTED:
+                logger.debug("Device {} CONNECTED", address);
+                execute();
+                break;
+            case DISCONNECTED:
+                logger.debug("Device {} DISCONNECTED", address);
+                break;
+            default:
+                break;
+        }
+    }
+
+    @Override
+    public void onCharacteristicReadComplete(BluetoothCharacteristic characteristic, BluetoothCompletionStatus status) {
+        try {
+            if (status == BluetoothCompletionStatus.SUCCESS) {
+                logger.debug("Characteristic {} from device {}: {}", characteristic.getUuid(), address,
+                        characteristic.getValue());
+                updateStatus(ThingStatus.ONLINE);
+                try {
+                    updateChannels(new AirthingsWavePlusDataParser(characteristic.getValue()));
+                } catch (AirthingsParserException e) {
+                    logger.warn("Data parsing error occured, when parsing data from device {}, cause {}", address,
+                            e.getMessage(), e);
+                }
+            } else {
+                logger.debug("Characteristic {} from device {} failed", characteristic.getUuid(), address);
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "No response from device");
```

Comment:
Wouldn't the Thing status almost immediately switch back to ONLINE when BeaconBluetoothHandler receives the next scanned rssi value?

Reply:
Sure, if scan result is received.